libs/utils/src/http/endpoint.rs

```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
+
+        tokio::task::spawn_blocking(move || {
+            let _span = span.entered();
+            // Currently we take a lot of mutexes while collecting metrics, so it's
+            // better to spawn a blocking task to avoid blocking the event loop.
+            let metrics = metrics::gather();
+            let mut buffer = Vec::new();
+            let encoder = TextEncoder::new();
+            encoder
+                .encode(&metrics, &mut buffer)
+                .map_err(|e| ApiError::InternalServerError(e.into()))?;
+
+            Ok(Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, encoder.format_type())
+                .body(Body::from(buffer))
+                .unwrap())
+        })
+        .await
+        .map_err(|e| ApiError::InternalServerError(e.into()))
+        .and_then(|e| e)
+    } else {
+        let (tx, rx) = mpsc::channel(1);
+
+        let body = Body::wrap_stream(ReceiverStream::new(rx));
+
+        struct ChannelWriter {
+            buffer: bytes::BytesMut,
+            tx: tokio::sync::mpsc::Sender<std::io::Result<bytes::Bytes>>,
+            written: usize,
+        }
 
-    let metrics = tokio::task::spawn_blocking(move || {
-        // Currently we take a lot of mutexes while collecting metrics, so it's
-        // better to spawn a blocking task to avoid blocking the event loop.
-        metrics::gather()
-    })
-    .await
-    .map_err(|e: JoinError| ApiError::InternalServerError(e.into()))?;
-    encoder.encode(&metrics, &mut buffer).unwrap();
+        impl ChannelWriter {
+            fn new(buf_len: usize, tx: mpsc::Sender<std::io::Result<Bytes>>) -> Self {
+                ChannelWriter {
+                    buffer: BytesMut::with_capacity(buf_len),
+                    tx,
+                    written: 0,
+                }
+            }
+
+            fn flush0(&mut self) -> std::io::Result<usize> {
+                let n = self.buffer.len();
+                if n > 0 {
+                    let ready = self.buffer.split().freeze();
+                    tracing::trace!(n = ready.len(), "flushing");
+
+                    // not ideal to call from blocking code to block_on, but we are sure that this
+                    // operation does not spawn_blocking other tasks
+                    let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
+                        self.tx.send(Ok(ready)).await.map_err(|_| ())?;
+
+                        // throttle sending to allow reuse of our buffer in `write`.
+                        self.tx.reserve().await.map_err(|_| ())?;
+
+                        // now the response task has picked up the buffer and hopefully started
+                        // sending it to the client.
+                        Ok(())
+                    });
+                    if res.is_err() {
+                        return Err(std::io::ErrorKind::BrokenPipe.into());
+                    }
```

Comment:
I can live with this, but why not a custom error telling that the receiver is gone?

If this error gets printed anywhere without enough context, people are going to start wondering "where are we even using a pipe".

Reply:
This is called within the `std::io::Write` impl. Very self-contained, not used from anywhere else than this one response writing out. It doesn't matter what the error is, but these plain `ErrorKind.into()` are the cheapest ones.